cmd/disk-cache.go

```diff
@@ -0,0 +1,894 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/djherbis/atime"
+
+	errors2 "github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/wildcard"
+
+	"github.com/minio/minio/pkg/hash"
+)
+
+// list of all errors that can be ignored in tree walk operation in disk cache
+var cacheTreeWalkIgnoredErrs = append(baseIgnoredErrs, errDiskAccessDenied, errVolumeNotFound, errFileNotFound)
+
+const (
+	// disk cache needs to have cacheSizeMultiplier * object size space free for a cache entry to be created.
+	cacheSizeMultiplier      = 100
+	cacheMaxDiskUsagePct     = 80 // in %
+	cacheFSObjectsLockPrefix = "cache"
+	cacheDataDir             = "data"
+)
+
+// abstract slice of cache drives backed by FS.
+type diskCache struct {
+	cfs []*cacheFSObjects
+}
+
+// Abstracts disk caching - used by the S3 layer
+type cacheObjects struct {
+	// pointer to disk cache
+	cache *diskCache
+	// ListObjects pool management.
+	listPool *treeWalkPool
+	// file path patterns to exclude from cache
+	exclude []string
+	// Object functions pointing to the corresponding functions of backend implementation.
+	GetObjectFn               func(bucket, object string, startOffset int64, length int64, writer io.Writer, etag string) (err error)
+	GetObjectInfoFn           func(bucket, object string) (objInfo ObjectInfo, err error)
+	PutObjectFn               func(bucket, object string, data *hash.Reader, metadata map[string]string) (objInfo ObjectInfo, err error)
+	ListObjectsFn             func(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error)
+	ListBucketsFn             func() (buckets []BucketInfo, err error)
+	GetBucketInfoFn           func(bucket string) (bucketInfo BucketInfo, err error)
+	NewMultipartUploadFn      func(bucket, object string, metadata map[string]string) (uploadID string, err error)
+	PutObjectPartFn           func(bucket, object, uploadID string, partID int, data *hash.Reader) (info PartInfo, err error)
+	AbortMultipartUploadFn    func(bucket, object, uploadID string) error
+	CompleteMultipartUploadFn func(bucket, object, uploadID string, uploadedParts []CompletePart) (objInfo ObjectInfo, err error)
+}
+
+// CacheObjectLayer implements primitives for cache object API layer.
+type CacheObjectLayer interface {
```

Comment:
why do we need a separate interface? we could simply leverage existing object layer interface.. and rest of the stub functions can have overrides.